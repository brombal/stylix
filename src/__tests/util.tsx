import '@testing-library/jest-dom';
import { cleanup, render } from '@testing-library/react';
import type React from 'react';

import { StylixProvider } from '../index';
import type { StylixMediaDefinition } from '../plugins/mediaObjects';

type RenderStylixReturn = readonly [string, string, ReturnType<typeof render>, HTMLStyleElement] & {
  containerHtml: string;
  styles: string;
  result: ReturnType<typeof render>;
  styleElement: HTMLStyleElement;
};

export function renderStylix(
  element: React.ReactElement,
  plugins?: any[],
  media?: StylixMediaDefinition,
): RenderStylixReturn {
  cleanup();
  const htmlStyleElement = document.createElement('style');
  document.head.appendChild(htmlStyleElement);
  const r = (
    <StylixProvider devMode styleElement={htmlStyleElement} plugins={plugins} media={media}>
      {element}
    </StylixProvider>
  );
  const result = render(r);

  const returnVal = [
    result.container.innerHTML,
    htmlStyleElement.innerHTML,
    result,
    htmlStyleElement,
  ] as unknown as RenderStylixReturn;
  returnVal.containerHtml = result.container.innerHTML;
  returnVal.styles = htmlStyleElement.innerHTML;
  returnVal.result = result;
  returnVal.styleElement = htmlStyleElement;
  return returnVal;
}
