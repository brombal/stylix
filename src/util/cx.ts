type ClassNamePrimitive = string | number | boolean | null | undefined;
type ClassName =
  | ClassNamePrimitive
  | ClassName[]
  | { [key: string]: ClassNamePrimitive }
  | (() => ClassName);

// Internal helper to collect class name parts without joining
function cxArray(args: ClassName[]): string[] {
  const classNames: string[] = [];
  for (const arg of args) {
    if (arg && typeof arg === 'string') {
      classNames.push(arg);
    } else if (Array.isArray(arg)) {
      classNames.push(...cxArray(arg));
    } else if (typeof arg === 'function') {
      classNames.push(...cxArray([arg()]));
    } else if (typeof arg === 'object' && arg !== null) {
      for (const [key, value] of Object.entries(arg)) {
        if (value) {
          classNames.push(key);
        }
      }
    }
  }
  return classNames;
}

/**
 * A utility function to create a string of class names based on the provided parameters.
 * Accepts a variable number of arguments, each of which can be one of the following:
 *
 * - A string, which will be included in the class name string.
 * - An object, where the keys are class names and the values are booleans indicating whether to include the class name.
 * - An array of strings or objects, which will be flattened and processed as above.
 * - A function that returns a string, object, or array, which will be processed as above.
 * - Any other value will be ignored.
 */
export function cx(...args: ClassName[]): string {
  return cxArray(args).join(' ');
}
