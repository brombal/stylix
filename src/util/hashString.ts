/**
 * Cheap string hashing, suitable for generating css class names
 */
export function hashString(str): string {
  let hash = 5381;
  let i = str.length;
  while (i) hash = (hash * 33) ^ str.charCodeAt(--i);
  return 'stylix-' + (hash >>> 0).toString(36);
}
