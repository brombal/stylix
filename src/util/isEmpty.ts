export function isEmpty(obj: unknown) {
  if (!obj) return true;
  if (Array.isArray(obj)) return obj.length === 0;
  for (const _ in obj) return false;
  if (typeof obj === 'object') return true;
  return false;
}
