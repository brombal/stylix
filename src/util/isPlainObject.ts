/**
 * Indicates that an object is most likely just an object literal.
 */
export function isPlainObject(value: any): value is Record<string, any> {
  if (!value || typeof value !== 'object') return false;
  return Object.getPrototypeOf(value) === Object.prototype;
}
