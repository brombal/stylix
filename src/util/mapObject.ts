export type MapObjectFunction = (
  key: string | number,
  value: any,
  source: unknown,
  context: any,
  mapRecursive: (value: unknown) => unknown,
) => unknown;

/**
 * Returns a new object or array, generated by invoking `map` on each key-value pair in `source` and merging the returned value into
 * the result. The return value should be an object or array (to match `source`), or undefined to omit the key-value pair from the result.
 *
 * If `source` is an object, Object.assign() will be used to merge the response into the result object.
 *
 * If `source` is an array, the returned array entries will be pushed onto the result array (i.e. it will be flattened, one level deep).
 *
 * The `map` function will receive the following arguments:
 * - The current key or index
 * - The current value
 * - The current object/array being mapped
 * - A context object. This is a plain object that you can modify as needed. The value will be shallow cloned for each key, and the clone
 *   will be passed into the `mapRecursive` method in order to persist it in recursive mapping.
 * - A `mapRecursive` function that can be used to recursively map nested objects or arrays. You can call this by passing the current value,
 *   which will run the same mapping function on the value and return the result. If the value is not an object or array, it will just
 *   return the value. The mapping function will receive the context object for the current key, and any modifications made to the context
 *   object will be persisted into the recursive call. This could be used, for example, to keep track of the current depth of the recursion
 *   or the full path of the current value.
 *
 * Example:
 *
 * ```ts
 * const value = {
 *   ignoreMe: null,
 *   string: 'value',
 *   object: {
 *     a: 1,
 *     b: 2,
 *   },
 *   array: [1, 2, 3, 4],
 * }
 * const result = mapObjectRecursive(value, (key, value, source, context) => {
 *   if (key === 'ignoreMe')
 *     return undefined; // will omit key from result
 *   if (typeof key === 'string' && typeof value === 'string')
 *     return { [key]: value + '-mapped' }; // will append '-mapped' to string values
 *   if (typeof key === 'string' && typeof value === 'number')
 *     return { [key]: value, [`${key}-mapped`]: value * 2 }; // will add a new key with the value doubled
 *   if (typeof value === 'object')
 *     return { [key]: value }; // will leave object unchanged
 *   if (Array.isArray(source) && typeof value === 'number')
 *     return [value, value * 2]; // will add the value and the value doubled to the array
 * });
 * // result:
 * // {
 * //   string: 'value-mapped',
 * //   object: {
 * //     a: 1,
 * //     'a-mapped': 2,
 * //     b: 2,
 * //     'b-mapped': 4,
 * //   },
 * //   array: [1, 2, 2, 4, 3, 6, 4, 8],
 * // }
 * ```
 */
export function mapObject<TSource>(
  source: TSource,
  map: MapObjectFunction,
  context: any = {},
): TSource {
  if (typeof source !== 'object') return source;
  const result = Array.isArray(source) ? ([] as unknown[]) : ({} as Record<string, unknown>);
  for (const _key in source) {
    const key = Array.isArray(source) ? +_key : _key;
    const value = (source as any)[key];

    const contextClone = { ...context };
    const mapResult = map(key, value, source, contextClone, (value) =>
      mapObject(value, map, contextClone),
    );

    if (typeof mapResult === 'undefined') continue;

    if (Array.isArray(mapResult) && Array.isArray(source)) {
      (result as unknown[]).push(...mapResult);
      continue;
    }

    if (
      typeof mapResult === 'object' &&
      !Array.isArray(mapResult) &&
      typeof source === 'object' &&
      !Array.isArray(source)
    ) {
      Object.assign(result as object, mapResult);
      continue;
    }

    throw new Error(
      `mapObjectRecursive: return value of map function must be an object, array, or undefined, and must match the type of the source value. Type of source value was ${typeof source}, and type of returned value was ${typeof mapResult}.`,
    );
  }

  return result as TSource;
}
