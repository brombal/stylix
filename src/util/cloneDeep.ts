import { isPlainObject } from './isPlainObject';

/**
 * Deeply clones a value.
 */
export function cloneDeep<T>(value: T): T {
  if (!value || typeof value !== 'object') return value;
  if (Array.isArray(value)) {
    const clone: any = [];
    for (let index = 0; index < value.length; ++index) {
      clone.push(cloneDeep(value[index]));
    }
    return clone;
  }
  if (isPlainObject(value)) {
    const clone: any = {};
    for (const key in value) {
      clone[key] = cloneDeep(value[key]);
    }
    return clone;
  }
  return value;
}
