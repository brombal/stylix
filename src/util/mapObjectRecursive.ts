import { cloneDeep } from './cloneDeep';
import { isPlainObject } from './isPlainObject';

/**
 * Invokes `map` on each key/value pair in `object`. The key/value pair is deleted from the object and replaced by
 * merging in the object returned from `map`. Recursively descends into all object and array values.
 * The `map` function will receive the key, the value, the current object being mapped, and a context object.
 * The context object is a plain object that you can modify as needed. The value will persist to subsequent calls to
 * `map` on child properties of `value`.
 */
export function mapObjectRecursive(
  object: any,
  map: (
    key: string | number,
    value: any,
    object: any,
    context: any,
  ) => Record<string | number, any>,
): any {
  const _mapObjectRecursive = function _mapObjectRecursive(
    object: any,
    map: (
      key: string | number,
      value: any,
      object: any,
      context: any,
    ) => Record<string | number, any>,
    context: any,
  ) {
    const clone = Array.isArray(object) ? [] : {};
    Object.entries(object).forEach(([key, value]: [number | string, any]) => {
      if (Array.isArray(object)) key = +key;
      const contextClone = cloneDeep(context);
      let result = map(key, value, object, contextClone);
      if (typeof result !== 'undefined' && typeof result !== 'object' && !Array.isArray(result))
        throw new Error(
          'mapObjectRecursive: return value of map function must be undefined, object, or array!',
        );
      if (typeof result === 'undefined') {
        result = { [key]: value };
      }
      Object.entries(result).forEach(([key, value]: [number | string, any]) => {
        if (isPlainObject(value) || Array.isArray(value))
          value = _mapObjectRecursive(value, map, contextClone);
        if (typeof value !== 'undefined') clone[key] = value;
      });
    });
    return clone;
  };
  return _mapObjectRecursive(object, map, {});
}
