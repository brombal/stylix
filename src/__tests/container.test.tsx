import $ from '../index';
import { renderStylix } from './util';

describe('container queries', () => {
  it('should render', () => {
    const el = (
      <$.div marginRight={4} container="parent / inline-size">
        <$.div
          $css={{
            '@container card (min-width: 1px)': {
              '&': {
                fontSize: '100cqw',
              },
            },
          }}
        />
      </$.div>
    );

    const [json, styles] = renderStylix(el);

    expect(json).toMatchSnapshot();
    expect(styles).toMatchSnapshot();
  });
});
