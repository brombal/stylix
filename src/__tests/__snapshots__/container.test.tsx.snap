// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`container queries should render 1`] = `"<div class="stylix-1"><div class="stylix-2"></div></div>"`;

exports[`container queries should render 2`] = `
".stylix-1 {
  margin-right: 4px;
  container: parent / inline-size; }
.stylix-2 {
@container card (min-width: 1px) {
& {
  font-size: 100cqw; } } }"
`;
